Support for GPU NUFFT

chaithyagr opened the pull request:
This PR Holds the support for GPU based NUFFT which is critical for 3D runs.

chaithyagr commented:
The drop in coverage is as we added a local test for GPU which we cant test on Travis, but we still to test locally whenever needed

LElgueddari commented on mri/reconstruct/fourier.py:
```diff
@@ -208,3 +215,227 @@ def adj_op(self, x):
+                               ft_axes=tuple(range(samples.shape[1])),
+                               radix=None)
+
+        elif self.platform == 'gpu':
```
You should consider excluding the GPU version from coverall. Check out [this](https://coverage.readthedocs.io/en/coverage-4.2/excluding.html) documentation.

chaithyagr replied:
As discussed, we decided to not do this as the goal is to not have 100 percent coverage but to know what parts arent tested. As these codes will be constantly used going forward, we plan to keep this ON. Maybe in future we could start running even the GPU versions of codes which could bump up the Coverage numbers

chaithyagr commented:
Rebased the codes, moved the code base of NFFT as NUFFT with platform as 'cpu'
Also, I have modified the tags as 'cuda' and 'opencl' than 'multi-cpu' and 'gpu' as it makes more sense.

Added 2 more local tests for NUFFT with opencl and they are working fine. 
Adding for re-review and merge.

chaithyagr commented:
@LElgueddari , I had a discussion with @zaccharieramzi  on the way we would want to structure the classes for NFFT. Here is what we felt is best implementation:

--- Fourier Base -> Ensures that the operator and adjoint Operator is implemented. 
        
--- NFFT(derived from Fourier_base) -> This is a base class for carrying out NFFT. This will hold some basic stuff which  is used by the implementations 
  -----> In this class, we will have a variable (say 'calculator') , which would hold an object of class which carries out implementation in CPU or on GPU with CUDA or OpenCL. **Note that this class will not be a parent class for NFFT.** We will initialize the right variable and do the same implementation based on what implementation we prefer.

Does this sound good?

I agree that this is cleanest implementation possible that follows the OOP concepts the right way.

chaithyagr commented:
Updated codes @LElgueddari , you can re-review and suggest any more changes. We will rebase this commit once #41 is at mainline

chaithyagr commented on mri/numerics/fourier.py:
```diff
@@ -13,5 +13,5 @@
 # Package import
 from mri.reconstruct.fourier import FFT2
-from mri.reconstruct.fourier import NFFT
+from mri.reconstruct.fourier import NUFFT
```
Yes, on it

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -26,6 +27,12 @@
+    from pynufft import NUFFT_hsa, NUFFT_cpu
+except Exception:
+    warnings.warn("pynufft python package has not been found. If needed use "
+                  "the master release. Till then you cant use NUFFT on GPU")
```
Both are same in my opinion, changing to cannot now..

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -208,3 +215,273 @@ def adj_op(self, x):
+            to (Jd,)*dims image
+        n_coils: int
+            Number of coils used to acquire the signal in case of multiarray
+            receiver coils
```
Done

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -208,3 +215,273 @@ def adj_op(self, x):
+                               Nd=self.shape,
+                               Kd=self.Kd,
+                               Jd=self.Jd,
+                               batch=1,  # self.nb_coils,
```
Yes thats the plan, will add TODO

chaithyagr commented:
Updated the codes and also rebased them so that the examples work fine. @LElgueddari can you please fix the build issue and then merge this PR so that we are sure that nothing is breaking

LElgueddari commented on examples/non_cartesian_reconstruction.py:
```diff
@@ -49,7 +49,7 @@
 
 # Get the locations of the kspace samples and the associated observations
-fourier_op = NFFT(samples=kspace_loc, shape=image.shape)
+fourier_op = NonCartesianFFT(samples=kspace_loc, shape=image.shape)
```
add the implementation parameter even if it is the default value

chaithyagr replied:
Sure

LElgueddari commented on mri/parallel_mri/extract_sensitivity_maps.py:
```diff
@@ -161,7 +161,7 @@ def get_Smaps(k_space, img_shape, samples, thresh,
             Smaps[l] = pfft.ifftshift(pfft.ifft2(pfft.fftshift(k_space[l])))
     elif mode == 'NFFT':
-        fourier_op = NFFT(samples=samples, shape=img_shape)
+        fourier_op = NonCartesianFFT(samples=samples, shape=img_shape)
```
add `implementation` parameters

LElgueddari commented on mri/reconstruct/fourier.py:
```diff
@@ -207,5 +231,284 @@ def adj_op(self, x):
+                                        n_coils=self.nb_coils)
+        else:
+            raise ValueError('Bad implementation ' + implementation +
+                             ' chosen')
```
I'm not sure that this is a good message error, I would recommend to add the different options inside message

chaithyagr replied:
Yes, will add options

LElgueddari commented on mri/test/test_fourier_adjoint.py:
```diff
@@ -26,13 +27,14 @@ def setUp(self):
         """
         self.N = 64
         self.max_iter = 10
+        self.num_channels = [1, 16]
```
For the test `num_channels = [1, 16]` is probably too much, instead, you could reduce it to `[1, 2]`

chaithyagr replied:
Sure.

chaithyagr commented:
> FFT2 should also be changed to take into account the number of coils to be more consistent with the rest of the code.
> FFT2 should be changed to FFTN to avoid duplicate code for he 3D case

Can we get this done when we are changing FFT2? ie in #45 ? I feel that is best place for this

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -189,8 +204,17 @@ def op(self, img):
+            coeff = np.asarray(coeff)
+        return coeff
+
+    def _adj_op(self, x):
```
Duplicate, resolving here. We will discuss this above.

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -206,5 +230,284 @@ def adj_op(self, x):
         """
-        self.plan.f = x
-        return np.copy(self.plan.adjoint()) / np.sqrt(self.plan.M)
+        if self.nb_coils == 1:
```
Ack.

chaithyagr replied:
I dont think I quite understand what needs to be added here.

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -206,5 +230,284 @@ def adj_op(self, x):
+                                                  'must be greater or equal ' \
+                                                  'than the image size'
+
+        print('Creating the NUFFT object...')
```
Yes Will print only if verbose=1

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -189,8 +204,17 @@ def op(self, img):
         """
-        self.plan.f_hat = img
-        return np.copy(self.plan.trafo()) / np.sqrt(self.plan.M)
+        if self.nb_coils == 1:
```
Same as below. Am I missing something here?

chaithyagr commented on mri/reconstruct/fourier.py:
```diff
@@ -170,10 +179,16 @@ def __init__(self, samples, shape):
         self.plan.precompute()
+        self.shape = shape
+
+    def _op(self, img):
```
Resolving for now, please reopen for discussion.

chaithyagr commented:
I made a few modifications for travis to ensure a passing build. @LElgueddari can you please merge?
